src/_pytest/logging.py

```diff
@@ -615,77 +619,63 @@ def pytest_collection(self) -> Generator[None, None, None]:
             else:
                 yield
 
-    @contextmanager
-    def _runtest_for(self, item, when):
-        with self._runtest_for_main(item, when):
+    def _runtest_for(self, item: nodes.Item, when: str) -> Generator[None, None, None]:
+        """Implements the internals of pytest_runtest_xxx() hook."""
+        if self.log_cli_handler is not None:
+            self.log_cli_handler.set_when(when)
+
+        with catching_logs(
+            handler=self._runtest_log_capture_handler,
+            formatter=self.formatter,
+            level=self.log_level,
+        ) as log_handler:
+            log_handler.reset()
+            item._store[catch_log_handler_key] = log_handler
+            item._store[catch_log_records_key][when] = log_handler.records
+
             if self.log_file_handler is not None:
                 with catching_logs(self.log_file_handler, level=self.log_file_level):
                     yield
             else:
                 yield
 
-    @contextmanager
-    def _runtest_for_main(
-        self, item: nodes.Item, when: str
-    ) -> Generator[None, None, None]:
-        """Implements the internals of pytest_runtest_xxx() hook."""
-        with catching_logs(
-            LogCaptureHandler(), formatter=self.formatter, level=self.log_level
-        ) as log_handler:
-            if self.log_cli_handler:
-                self.log_cli_handler.set_when(when)
-
-            if item is None:
-                yield  # run the test
-                return
-
-            if not hasattr(item, "catch_log_handlers"):
-                item.catch_log_handlers = {}  # type: ignore[attr-defined]
-            item.catch_log_handlers[when] = log_handler  # type: ignore[attr-defined]
-            item.catch_log_handler = log_handler  # type: ignore[attr-defined]
-            try:
-                yield  # run test
-            finally:
-                if when == "teardown":
-                    del item.catch_log_handler  # type: ignore[attr-defined]
-                    del item.catch_log_handlers  # type: ignore[attr-defined]
-
             if self.print_logs:
                 # Add a captured log section to the report.
                 log = log_handler.stream.getvalue().strip()
                 item.add_report_section(when, "log", log)
 
     @pytest.hookimpl(hookwrapper=True)
     def pytest_runtest_setup(self, item):
-        with self._runtest_for(item, "setup"):
-            yield
+        empty = {}  # type: Dict[str, List[logging.LogRecord]]
```

Comment:
Maybe you can split up the type definiton of `catch_log_records_key` into two:
* `catch_log_records_map`
* `catch_log_records_key`

where the latter is defined as `Store[catch_log_records_key]`. Thus you can use `empty = {} # type catch_log_records_map`, here. TBH I don't know if this really works

Reply:
These I think will also go away eventually. They are used for the plugin<->fixture interaction, but I think we should refactor this to be more like what the capture plugin does (for #7162).